### PR TITLE
[DO NOT MERGE] [THREESCALE-146] Overrides host and adds auth params in Swagger spec

### PR DESCRIPTION
**Pull request by josemigallas**
⚠️ Do not merge or review this PR

**Which issue(s) this PR fixes** 

[THREESCALE-1460 - Improve import of OAS in ActiveDocs](https://issues.jboss.org/browse/THREESCALE-1460)

**Verification steps** 

1. Select any service
2. Go to **Integration > Configuration** and take notice of what's the **Production Public Base URL**
3. Go to **ActiveDocs** and select an existing spec or create one from scratch (you may use [SwaggerUI's example](https://petstore.swagger.io/v2/swagger.json))
4. Take notice of what's in the JSON's `host` field (`petstore.swagger.io` from the aforementioned example), even change it for something made up
5. Go back to the actual spec and try any request: verify the endpoint is the one saw at (1) regardless of what was defined in the JSON.
6. Go to **Integration > Configuration > Edit integration settings** and select **API Key** as Authentication method.
7. Go to **ActiveDocs** and select any request: it should have an API Key parameter.
8. Go to **Integration > Configuration > Edit integration settings** and select **App_ID and App_KeyPair** as Authentication method.
9. Go to **ActiveDocs** and select any request: it should have an App ID and an App Key parameters.

**For Discussion**
The solution for this issue is based on an _on-the-fly_ modification of the Swagger spec that's retrieved from the rails application. This modification is performed by Javascript, that is by the user's browser.

There's however the possibility to do this from the very Rails application, preventing a possible slow performance. The modifications are just updating the JSON before feeding it to SwaggerUI so it doesn't really matter where this is done.

The question is:

- do we cope with the workload in our own machines, beeing able to scalate and improving UX but also by having more workload?
- do we sent all workload to the user which will have to handle possibly a huge JSON and iterate over its params and do changes all within the browser's memory (hard for slow machines, old browsers, mobile devices..)?

### CODE REVIEW

File: lib/developer_portal/lib/liquid/tags/active_docs.rb [hunk 1]
```diff
@@ -21,14 +21,16 @@ def initialize(tag_name, params, tokens)
...
-            context.registers[:controller].send(:site_account)
-            .api_docs_services.published
-            .select{|e| e.specification.swagger_2_0?}.map(&:system_name)
+          @services = api_docs_services.select{|e| e.specification.swagger_2_0?}.map(&:system_name)
         end
 
-        render_erb context, "shared/#{version}", {services: @services}
+        @api_docs_services = api_docs_services.where(system_name: @services)
+
+        render_erb context, "shared/#{version}", {services: @services, api_docs_services: @api_docs_services}
```
Comment (hallelujah) [inline lib/developer_portal/lib/liquid/tags/active_docs.rb]: why do you need that?

```liquid
{% for service in services %}
  {% for spec in service.api_specs %}
  // do whatever
  {% endfor %}
{% endfor %}

```

Comment (hallelujah) [reply lib/developer_portal/lib/liquid/tags/active_docs.rb]: There is also https://github.com/3scale/porta/blob/5b9187805193e4bb8b696c21df9edb6af1e7889a/lib/developer_portal/lib/liquid/drops/provider.rb#L202-L212

`provider` is available everywhere

File: config/locales/en.yml [hunk 2]
```diff
@@ -1095,7 +1095,7 @@ en:
     credentials_location:
       headers: "As HTTP Headers"
       query: "As query parameters (GET) or body parameters (POST/PUT/DELETE)"
-      authorization: "As HTTP Basic Authorization"
+      authorization: "As HTTP Basic Authentication"
```
Comment (hallelujah) [inline config/locales/en.yml]: Belongs to another PR right?

File: features/provider/integration/edit.feature [hunk 3]
```diff
@@ -19,7 +19,7 @@ Feature: Edit Integration
     Given the service uses app_id/app_key as authentication method
     And I go to the service integration page
     And I toggle "Authentication Settings"
-    When I choose "As HTTP Basic Authorization"
+    When I choose "As HTTP Basic Authentication"
```
Comment (hallelujah) [inline features/provider/integration/edit.feature]: Belongs to another PR I think

File: lib/developer_portal/lib/liquid/tags/active_docs.rb (hunk 1 above)
Comment (guicassolato) [reply lib/developer_portal/lib/liquid/tags/active_docs.rb]: @hallelujah, I think `services` is an array of service system names.

Comment (hallelujah) [reply lib/developer_portal/lib/liquid/tags/active_docs.rb]: Oh services are the api_spec system_name!!! 🤦‍♂️ 
Anyway I do not think we should add it here. You can always use  `provider.api_specs` and filter

Comment (josemigallas) [reply lib/developer_portal/lib/liquid/tags/active_docs.rb]: @guicassolato added this to make `@api_docs_services` available from **_swagger_*.html.erb**, so that we have access to `host_with_port`. I don't understand what your comments have to do with this.

File: features/provider/integration/edit.feature (hunk 3 above)
Comment (josemigallas) [reply features/provider/integration/edit.feature]: Certainly...

File: config/locales/en.yml (hunk 2 above)
Comment (josemigallas) [reply config/locales/en.yml]: Certainly!!

File: app/views/admin/api_docs/base/swagger.html.erb [hunk 4]
```diff
@@ -37,18 +37,70 @@
       }
     }
 
+    function injectAuthParams(spec) {
+      var version = 2; // <%= @api_docs_service.service.proxy.authentication_method %>;
```
Comment (hallelujah) [inline app/views/admin/api_docs/base/swagger.html.erb]: Uncommenting this?

File: lib/developer_portal/lib/liquid/drops/api_spec.rb [hunk 5]
```diff
@@ -21,6 +21,11 @@ def service
         return unless (service = @spec.service)
         Drops::Service.new(service)
       end
+
+      desc 'Returns the public production host endpoint if it linked to a service `nil` otherwise'
```
Comment (didierofrivia) [inline lib/developer_portal/lib/liquid/drops/api_spec.rb]: 'Returns the public production host endpoint if **it's** linked to a service**;** `nil` otherwise'

File: lib/developer_portal/lib/liquid/tags/active_docs.rb [hunk 6]
```diff
@@ -43,6 +45,9 @@ def version
         end
       end
 
+      def swagger2?
+        @version > "1.2"
```
Comment (didierofrivia) [inline lib/developer_portal/lib/liquid/tags/active_docs.rb]: we probably would need to make sure if there's anyone still using version < 2.

**Comment by didierofrivia** [comment]
As I commented earlier, I'd rather do the modification with a presenter/decorator in rails than in js. Would be faster and cleaner than doing it in the browser.

File: lib/developer_portal/lib/liquid/tags/active_docs.rb (hunk 6 above)
Comment (josemigallas) [reply lib/developer_portal/lib/liquid/tags/active_docs.rb]: How do we do that?

File: app/views/admin/api_docs/base/swagger.html.erb [hunk 7]
```diff
@@ -37,18 +37,70 @@
...
+      var version = <%= @api_docs_service.service.proxy.authentication_method %>;
+      var authParams;
+      
+      switch(version) {
+        case 2:
+          authParams = [{
+            name: 'x-3scale-auth-app-key',
+            in: 'header',
+            description: 'App Key authentication parameter',
+            'x-data-three-scale-name': 'app_key',
```
Comment (hallelujah) [inline app/views/admin/api_docs/base/swagger.html.erb]: This should be `app_keys`
https://access.redhat.com/documentation/en-us/red_hat_3scale/2.3/html/api_documentation/create-activedocs-spec

File: app/views/admin/api_docs/base/swagger.html.erb [hunk 8]
```diff
@@ -37,18 +37,70 @@
...
+            required: true,
+            type: 'string'
+          },{
+            name: 'x-3scale-auth-app-id',
+            in: 'header',
+            description: 'App ID authentication parameter',
+            'x-data-three-scale-name': 'app_id',
```
Comment (hallelujah) [inline app/views/admin/api_docs/base/swagger.html.erb]: This should be `app_ids`

File: app/views/admin/api_docs/base/swagger.html.erb [hunk 9]
```diff
@@ -37,18 +37,70 @@
...
+            type: 'string'
+            }]
+          break;
+        
+        case 1:   
+          authParams = [{
+            name: 'x-3scale-auth-user-key',
+            in: 'header',
+            description: 'User Key authentication parameter',
+            'x-data-three-scale-name': 'user_key',
```
Comment (hallelujah) [inline app/views/admin/api_docs/base/swagger.html.erb]: This should be `user_keys`

File: app/views/admin/api_docs/base/swagger.html.erb [hunk 10]
```diff
@@ -37,18 +37,70 @@
...
     }
 
+    function injectAuthParams(spec) {
+      var version = <%= @api_docs_service.service.proxy.authentication_method %>;
+      var authParams;
+      
+      switch(version) {
+        case 2:
+          authParams = [{
+            name: 'x-3scale-auth-app-key',
```
Comment (hallelujah) [inline app/views/admin/api_docs/base/swagger.html.erb]: This should be `<%= @api_docs_service.service.proxy.auth_app_key %>`

File: app/views/admin/api_docs/base/swagger.html.erb [hunk 11]
```diff
@@ -37,18 +37,70 @@
...
+            in: 'header',
+            description: 'App Key authentication parameter',
+            'x-data-three-scale-name': 'app_key',
+            required: true,
+            type: 'string'
+          },{
+            name: 'x-3scale-auth-app-id',
```
Comment (hallelujah) [inline app/views/admin/api_docs/base/swagger.html.erb]: This should be `<%= @api_docs_service.service.proxy.auth_app_id %>`

File: app/views/admin/api_docs/base/swagger.html.erb [hunk 12]
```diff
@@ -37,18 +37,70 @@
...
+            description: 'App ID authentication parameter',
+            'x-data-three-scale-name': 'app_id',
+            required: true,
+            type: 'string'
+            }]
+          break;
+        
+        case 1:   
+          authParams = [{
+            name: 'x-3scale-auth-user-key',
```
Comment (hallelujah) [inline app/views/admin/api_docs/base/swagger.html.erb]: This should be `<%= @api_docs_service.service.proxy.auth_user_key %>`

File: app/views/admin/api_docs/base/swagger.html.erb [hunk 13]
```diff
@@ -37,18 +37,70 @@
...
+          break;
+            
+        default:
+        return;
+      }
+          
+      for (var p in spec.paths) {
+        var path = spec.paths[p];
+        for (var verb in path) {
+          Array.prototype.push.apply(path[verb].parameters, authParams);
```
Comment (hallelujah) [inline app/views/admin/api_docs/base/swagger.html.erb]: we should check that the parameter is not inside the spec already otherwise it will be added twice

File: lib/developer_portal/lib/liquid/tags/active_docs.rb [hunk 14]
```diff
@@ -16,19 +16,21 @@ def initialize(tag_name, params, tokens)
         super
         if params =~ Syntax
           @version  = $1.strip
-          @services = $2.nil? ? [] : $2.split(",").map(&:strip)
+          @system_names = $2.nil? ? [] : $2.split(",").map(&:strip)
         end
       end
 
+      attr_reader :system_names
```
Comment (hallelujah) [inline lib/developer_portal/lib/liquid/tags/active_docs.rb]: This will be public? So they can use it in the liquid template?

File: lib/developer_portal/lib/liquid/tags/active_docs.rb (hunk 6 above)
Comment (hallelujah) [reply lib/developer_portal/lib/liquid/tags/active_docs.rb]: It should be part of deprecation plan for the road to OAS 3?
/cc @3scale/product

**Comment by hallelujah** [comment]
@didierofrivia yes it is as easy and we found out that there are 2 controllers that uses it

https://github.com/3scale/porta/blob/265948583efcabaaf5d2fb791ac202f42047527c/app/controllers/admin/api_docs/base_controller.rb

https://github.com/3scale/porta/blob/265948583efcabaaf5d2fb791ac202f42047527c/lib/developer_portal/app/controllers/developer_portal/swagger/spec_controller.rb

File: app/views/admin/api_docs/base/swagger.html.erb (hunk 13 above)
Comment (josemigallas) [reply app/views/admin/api_docs/base/swagger.html.erb]: Very good point.

File: lib/developer_portal/lib/liquid/tags/active_docs.rb (hunk 14 above)
Comment (guicassolato) [reply lib/developer_portal/lib/liquid/tags/active_docs.rb]: I didn't think this through. We can remove that.

Comment (guicassolato) [reply lib/developer_portal/lib/liquid/tags/active_docs.rb]: Done

File: app/views/admin/api_docs/base/swagger.html.erb (hunk 13 above)
Comment (didierofrivia) [reply app/views/admin/api_docs/base/swagger.html.erb]: and this also could be in a decorator, right?

File: doc/liquid/drops.md [hunk 15]
```diff
@@ -365,6 +365,9 @@ Returns the name of the spec.
 ### service
 Returns the service of the spec if it has any or `nil` otherwise.
 
+### host_with_port
+Returns the public production host endpoint if it linked to a service `nil` otherwise
```
Comment (guicassolato) [inline doc/liquid/drops.md]: > [...] if it **is** linked to a service **or** ...

File: test/unit/liquid/drops/api_spec_drop_test.rb [hunk 16]
```diff
@@ -26,6 +26,11 @@ def setup
     assert_nil api_spec.service
   end
 
+  test 'returns the host with port' do
+    @spec.expects(host_with_port: 'example.com:3000')
+    assert_equal 'example.com:3000', api_spec.host_with_port
+  end
```
Comment (guicassolato) [inline test/unit/liquid/drops/api_spec_drop_test.rb]: Can we add a test for cases where
1. the port is not part of the proxy endpoint
2. there's is no service linked to the apidoc
?

File: app/views/admin/api_docs/base/swagger.html.erb [hunk 17]
```diff
@@ -37,18 +37,80 @@
       }
     }
 
+    var AUTH_METHOD_USER_KEY = 1;
```
Comment (guicassolato) [inline app/views/admin/api_docs/base/swagger.html.erb]: `const`?

File: app/views/admin/api_docs/base/swagger.html.erb [hunk 18]
```diff
@@ -99,16 +161,23 @@
 
   <script type="text/javascript">
     $(function () {
-      var url = "<%= preview_admin_api_docs_service_path(@api_docs_service, format: :json)-%>";
+      var host = "<%= @api_docs_service.host_with_port %>";
+      var spec = <%= ActionController::Base.helpers.sanitize(@api_docs_service.body) %>;
```
Comment (guicassolato) [inline app/views/admin/api_docs/base/swagger.html.erb]: It should work only with `sanitize @api_docs_service.body`

File: app/models/api_docs/service.rb [hunk 19]
```diff
@@ -27,6 +27,7 @@ class ApiDocs::Service < ApplicationRecord
   scope :published, -> { where(published: true) }
   scope :accessible, -> { joining { service.outer }.where.has { (service_id == nil) | (service.state != ::Service::DELETE_STATE) } }
   scope :without_service, -> { where(service_id: nil) }
+  scope :swagger2, -> { where('swagger_version >= ?', '2.0') }
```
Comment (Martouta) [inline app/models/api_docs/service.rb]:
```suggestion
  scope :swagger2, -> { where.has { swagger_version >= '2.0' } }
```
BabySqueel style for `>=` 😄

File: app/views/admin/api_docs/base/swagger.html.erb [hunk 20]
```diff
@@ -99,16 +161,23 @@
...
+
+      injectAuthParams(spec);
+
       window.swaggerUi = new SwaggerUi({
-        url: url,
+        spec: spec,
```
Comment (guicassolato) [inline app/views/admin/api_docs/base/swagger.html.erb]: Is it possible to refactor `injectAuthParams` so, instead of changing the state, it actually returns a new `spec` and then we replace this with `spec: injectAuthParams(spec)`?

File: app/models/api_docs/service.rb [hunk 21]
```diff
@@ -65,6 +66,13 @@ def base_path
     specification.base_path || self[:base_path]
   end
 
+  def host_with_port
```
Comment (Martouta) [inline app/models/api_docs/service.rb]: It seems that is only for presentation logic stuff, so perhaps [a presenter](https://github.com/3scale/porta/tree/master/app/presenters) or [a decorator](https://github.com/3scale/porta/tree/master/app/decorators) containing this method instead of putting it in the model.

File: app/views/admin/api_docs/base/swagger.html.erb (hunk 17 above)
Comment (Martouta) [reply app/views/admin/api_docs/base/swagger.html.erb]: I am very confused about this `var` vs `const` or `let` in our case. We removed all ES6/7 stuff for capybara, right? But since we don't use Jenkins anymore, we can come back to add ES6/7 stuff, right? @josemigallas @damianpm @didierofrivia @guicassolato

File: lib/developer_portal/app/views/developer_portal/shared/_swagger_ui.html.liquid [hunk 22]
```diff
@@ -8,17 +8,79 @@
 </div>
 
 <script type="text/javascript">
+  var AUTH_METHOD_USER_KEY = 1;
+  var AUTH_METHOD_ID_KEY_PAIR = 2;
+
+  function getAuthParamsForMethod(authMethod) {
```
Comment (guicassolato) [inline lib/developer_portal/app/views/developer_portal/shared/_swagger_ui.html.liquid]: I don't think this is the kind of logic a provider would want to modify. Maybe we could move to its own js file and include it here. This would also avoid code duplication in `app/views/admin/api_docs/base/swagger.html.erb`. WDYT?

File: lib/developer_portal/lib/liquid/tags/active_docs.rb [hunk 23]
```diff
@@ -16,19 +16,19 @@ def initialize(tag_name, params, tokens)
         super
         if params =~ Syntax
           @version  = $1.strip
-          @services = $2.nil? ? [] : $2.split(",").map(&:strip)
+          @system_names = $2.nil? ? [] : $2.split(",").map(&:strip)
```
Comment (Martouta) [inline lib/developer_portal/lib/liquid/tags/active_docs.rb]: I personally advocate for:
```suggestion
          @system_names = ($2 || '').split(",").map(&:strip)
```
Because the only reason is doing the `nil` check is to don't break when it does `split` because it expects a String. So this way you send a String as it expects and if `$2` contains `nil`, you end up returning `[]` as well 😄

File: lib/developer_portal/lib/liquid/tags/active_docs.rb [hunk 24]
```diff
@@ -16,19 +16,19 @@ def initialize(tag_name, params, tokens)
...
         end
       end
 
       def render context
-        if @services.empty? && @version > "1.2"
-          @services =
-            context.registers[:controller].send(:site_account)
-            .api_docs_services.published
-            .select{|e| e.specification.swagger_2_0?}.map(&:system_name)
+        provider_account = context.registers[:controller].send(:site_account)
```
Comment (Martouta) [inline lib/developer_portal/lib/liquid/tags/active_docs.rb]: I know this is not your fault, but if `site_account` is private, I feel that we are doing something not really cool... 😞 Maybe it should be public and call it directly in a normal way... but I am not sure. WDYT @josemigallas @hallelujah @guicassolato @macejmic ?

File: test/unit/liquid/drops/api_spec_drop_test.rb (hunk 16 above)
Comment (Martouta) [reply test/unit/liquid/drops/api_spec_drop_test.rb]: I agree with Gui about this 👍

File: lib/developer_portal/lib/liquid/tags/active_docs.rb (hunk 6 above)
Comment (Martouta) [reply lib/developer_portal/lib/liquid/tags/active_docs.rb]: @josemigallas In case that we need to do that, you will need to connect to the preview DB going to [the deploy repository](https://github.com/3scale/deploy) locally and doing:
```
./bin/remote system preview01 ssh // connect to preview
RAILS_ENV=preview bundle exec rails console // use rails console inside preview
```
And then you can do: `ApiDocs::Service.where.has { swagger_version < '2.0' }.count` to see how many we have 😄 And this is the result at the moment:
image

Comment (Martouta) [reply lib/developer_portal/lib/liquid/tags/active_docs.rb]: But this is the amount of services, not the amount of accounts, which will be less but still...

File: app/views/admin/api_docs/base/swagger.html.erb (hunk 17 above)
Comment (josemigallas) [reply app/views/admin/api_docs/base/swagger.html.erb]: This file uses vanilla JS (no fancy modern syntax), we let ES6 features for `.jsx` files since they are transpiled by babel.

File: app/views/admin/api_docs/base/swagger.html.erb (hunk 18 above)
Comment (josemigallas) [reply app/views/admin/api_docs/base/swagger.html.erb]: What do you mean?

File: app/views/admin/api_docs/base/swagger.html.erb (hunk 20 above)
Comment (josemigallas) [reply app/views/admin/api_docs/base/swagger.html.erb]: Yes it is, but is there any particular reason?

File: app/views/admin/api_docs/base/swagger.html.erb (hunk 18 above)
Comment (Martouta) [reply app/views/admin/api_docs/base/swagger.html.erb]: He means like this I think:
```suggestion
      var spec = <%= sanitize @api_docs_service.body %>;
```
My bad btw, I suggested it like this without realising about this detail.

File: lib/developer_portal/app/views/developer_portal/shared/_swagger_ui.html.liquid (hunk 22 above)
Comment (josemigallas) [reply lib/developer_portal/app/views/developer_portal/shared/_swagger_ui.html.liquid]: Sounds like the proper approach

File: test/unit/liquid/drops/api_spec_drop_test.rb (hunk 16 above)
Comment (josemigallas) [reply test/unit/liquid/drops/api_spec_drop_test.rb]: >     1. the port is not part of the proxy endpoint

Meaning, the port is undefined or host_with_port = 'example.com'

>    2. there's is no service linked to the apidoc

How can I do that?

Comment (Martouta) [reply test/unit/liquid/drops/api_spec_drop_test.rb]: > 1. there's is no service linked to the apidoc

`spec = FactoryBot.build(:api_docs_service, service: nil)`
@josemigallas I am not sure if this solves your doubt 😄

File: app/views/admin/api_docs/base/swagger.html.erb (hunk 17 above)
Comment (Martouta) [inline app/views/admin/api_docs/base/swagger.html.erb]: Personally I am not a big fan of populating the global scope with constants or stuff that we won't need there. So I would put all this new stuff inside the anonymous function before (the one currently in line 96)

File: app/views/admin/api_docs/base/swagger.html.erb [hunk 25]
```diff
@@ -37,18 +37,80 @@
       }
     }
 
+    var AUTH_METHOD_USER_KEY = 1;
+    var AUTH_METHOD_ID_KEY_PAIR = 2;
+
+    function getAuthParamsForMethod(authMethod) {
```
Comment (Martouta) [inline app/views/admin/api_docs/base/swagger.html.erb]: There could be less repetition here, for example like this (there are more options):
```js
function getAuthParamsForMethod(authMethod) {
  if(authMethod !== AUTH_METHOD_ID_KEY_PAIR && authMethod !== AUTH_METHOD_USER_KEY) { return []; }
  
  var authParam = {
    name: '<%= @api_docs_service.service.proxy.auth_app_key %>',
    in: 'header',
    'x-data-threescale-name': 'app_keys',
    required: true,
    type: 'string'
  };
  
  if(authMethod === AUTH_METHOD_USER_KEY) { return [Object.assign(authParam, {description: 'User Key authentication parameter'})] }
  return [
    Object.assign(authParam, {description: 'App Key authentication parameter'}),
    {
      name: '<%= @api_docs_service.service.proxy.auth_app_id %>',
      in: 'header',
      description: 'App ID authentication parameter',
      'x-data-threescale-name': 'app_ids',
      required: true,
      type: 'string'
    }
  ];
}
```

File: lib/developer_portal/app/views/developer_portal/shared/_swagger_ui.html.liquid [hunk 26]
```diff
@@ -8,17 +8,79 @@
...
+          }
+        });
+      }
+    }
+  }
+
   (function () {
-    var url = "{{provider.api_specs.first.url}}";
+    var host = "<%= @api_docs_service.host_with_port %>";
+    var spec = <%= ActionController::Base.helpers.sanitize(@api_docs_service.body) %>;
```
Comment (Martouta) [inline lib/developer_portal/app/views/developer_portal/shared/_swagger_ui.html.liquid]: Same as https://github.com/3scale/porta/pull/539/files#r257965975
Whatever we decide for there, we should do the same here.

File: lib/developer_portal/app/views/developer_portal/shared/_swagger_ui_https.html.liquid [hunk 27]
```diff
@@ -8,17 +8,79 @@
...
+          }
+        });
+      }
+    }
+  }
+
   (function () {
-    var url = "{{provider.api_specs.first.url}}";
+    var host = "<%= @api_docs_service.host_with_port %>";
+    var spec = <%= ActionController::Base.helpers.sanitize(@api_docs_service.body) %>;
```
Comment (Martouta) [inline lib/developer_portal/app/views/developer_portal/shared/_swagger_ui_https.html.liquid]: Same as https://github.com/3scale/porta/pull/539/files#r257965975
Whatever we decide for there, we should do the same here.

File: lib/developer_portal/app/views/shared/_swagger.html.erb [hunk 28]
```diff
@@ -13,15 +13,24 @@
   <p><a href='#'>Sign in</a> to your account for quick access to useful values.</p>
 </div>
 
+<% url_opts = { format: :json }; url_opts[:services] = services.pluck(:system_name).join(',') unless services.empty? %>
+
 <script type="text/javascript">
   $(function () {
-    var url = "<%= services.empty? ? swagger_spec_index_url(format: :json) : swagger_spec_index_url(format: :json, services: services.join(',')) %>";
+    var host = "<%= @api_docs_service.host_with_port %>";
+    var spec = <%= ActionController::Base.helpers.sanitize(@api_docs_service.body) %>;
```
Comment (Martouta) [inline lib/developer_portal/app/views/shared/_swagger.html.erb]: Same as https://github.com/3scale/porta/pull/539/files#r257965975
Whatever we decide for there, we should do the same here.

File: lib/developer_portal/app/views/shared/_swagger_2_0.html.erb [hunk 29]
```diff
@@ -23,16 +23,79 @@
...
+          }
+        });
+      }
+    }
+  }
+
   $(function () {
-    var url = "<%= swagger_spec_url(services.first, format: :json) -%>";
+    var host = "<%= @api_docs_service.host_with_port %>";
+    var spec = <%= ActionController::Base.helpers.sanitize(@api_docs_service.body) %>;
```
Comment (Martouta) [inline lib/developer_portal/app/views/shared/_swagger_2_0.html.erb]: Same as https://github.com/3scale/porta/pull/539/files#r257965975
Whatever we decide for there, we should do the same here.

File: app/views/admin/api_docs/base/swagger.html.erb [hunk 30]
```diff
@@ -37,18 +37,80 @@
...
+            }
+          });
+        }
+      }
+    }
+
     $(function () {
-      var url = "<%= admin_api_docs_service_path(@api_docs_service, format: :json)-%>";
+      var host = "<%= @api_docs_service.host_with_port %>";
+      var spec = <%= ActionController::Base.helpers.sanitize(@api_docs_service.body) %>;
```
Comment (Martouta) [inline app/views/admin/api_docs/base/swagger.html.erb]: Same as https://github.com/3scale/porta/pull/539/files#r257965975
Whatever we decide for there, we should do the same here.

File: lib/developer_portal/app/views/developer_portal/shared/_swagger_ui.html.liquid (hunk 22 above)
Comment (Martouta) [inline lib/developer_portal/app/views/developer_portal/shared/_swagger_ui.html.liquid]: https://github.com/3scale/porta/pull/539/files#r258106136
Is this method exactly the same? Then we should probably have it only once if possible 🤔

File: lib/developer_portal/app/views/developer_portal/shared/_swagger_ui_https.html.liquid [hunk 31]
```diff
@@ -8,17 +8,79 @@
 </div>
 
 <script type="text/javascript">
+  var AUTH_METHOD_USER_KEY = 1;
+  var AUTH_METHOD_ID_KEY_PAIR = 2;
+
+  function getAuthParamsForMethod(authMethod) {
```
Comment (Martouta) [inline lib/developer_portal/app/views/developer_portal/shared/_swagger_ui_https.html.liquid]: Same as https://github.com/3scale/porta/pull/539/files#r258131749 🤔

File: lib/developer_portal/app/views/shared/_swagger_2_0.html.erb [hunk 32]
```diff
@@ -23,16 +23,79 @@
     }
   }
 
+  var AUTH_METHOD_USER_KEY = 1;
+  var AUTH_METHOD_ID_KEY_PAIR = 2;
+
+  function getAuthParamsForMethod(authMethod) {
```
Comment (Martouta) [inline lib/developer_portal/app/views/shared/_swagger_2_0.html.erb]: Same as https://github.com/3scale/porta/pull/539/files#r258131749 🤔

File: app/views/admin/api_docs/base/swagger.html.erb (hunk 20 above)
Comment (guicassolato) [reply app/views/admin/api_docs/base/swagger.html.erb]: It's always better when you know your function does not change the state IMO and in this case it's not needed to anyway.

Comment (josemigallas) [reply app/views/admin/api_docs/base/swagger.html.erb]: The problem is that this is all happening in the browser, so we are not only filling users's memory with the whole JSON once but we want to do a deep copy of it in order not to modify it.

You definitely have a point here, I just don't think it is worth it in this particular case.

We could actually argue we are even changing the app's state by doing this. We are retrieving an object, we are working with it and finally using it. There is no state involved.

File: app/models/api_docs/service.rb (hunk 21 above)
Comment (josemigallas) [reply app/models/api_docs/service.rb]: That sounds like a rather simple change, would you kindly create a commit suggestion?

Comment (Martouta) [reply app/models/api_docs/service.rb]: sure 😉 I will 👍

Comment (Martouta) [reply app/models/api_docs/service.rb]: @josemigallas https://github.com/3scale/porta/pull/621 😄

**Comment by thomasmaas** [comment]
I'm closing this @josemigallas, I think we agree we chose the wrong implementation. Doing it in ruby will be much dryer in this case.
